readme.md
# flo

flo is a command line application that keeps you in check. It’s small and fast.

## installation

    make
    cp flo /usr/local/bin/

## getting started

### add todo

     $ flo clean apartment
       0  clean apartment

### add event

     $ flo eat lunch@some cafe,221200-221245
       0  2010-07-22 12:00  eat lunch@some cafe
          2010-07-22 12:45
       1  clean apartment

### add deadline

     $ flo a deadline@uni-10011200
       0  2010-07-22 12:00  eat lunch@some cafe
          2010-07-22 12:45
    d  1  2010-10-01 12:00  a deadline
       2  clean apartment

### change item

     $ flo -c 0 -f 221300 -t 221345
       0  2010-07-22 13:00  eat lunch@some cafe
          2010-07-22 13:45
    d  1  2010-10-01 12:00  a deadline
       2  clean apartment

### remove item

     $ flo -r 2
       0  2010-07-22 13:00  eat lunch@some cafe
          2010-07-22 13:45
    d  1  2010-10-01 12:00  a deadline

## usage

    flo [what[@at][,from][-to] || [-c id] -w what | -a at | -f from | -t to || -r id]

### date formats

    YYMMDDhhmm
      MMDDhhmm
        DDhhmm
        DDhh
        DD

If the year or the month isn’t specified, the current year and month is used.
For formats without a month, if the date specified is before today’s date, the
month is set to the next month.

The value for hours and minutes is set to `00` if no other value is specified.

## license

ISC.


changelog
2010-07-18 try to print time_t correctly; cast it to an unsigned long long (Simon Zimmermann)
2010-07-18 Add warnings to compiler option(-Wall) and changelog(log) to Makefile (Simon Zimmermann)
2010-07-18 fix to date formatting (Alexander Teinum)
2010-07-18 add deadline functionality (Alexander Teinum)
2010-07-18 add digits that will later be used for deleting items (Alexander Teinum)
2010-07-18 sort items (Alexander Teinum)
2010-07-18 remove .f.c.swp, add .gitignore (Alexander Teinum)
2010-07-18 clean up (Alexander Teinum)
2010-07-18 rename readme.mk to readme.md (Alexander Teinum)
2010-07-18 first commit (Alexander Teinum)

f.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define LIST_ITEMS 128
#define _IS_DEADLINE(it) (it.from == 0 && it.to != 0)
#define IS_DEADLINE(it) (it->from == 0 && it->to != 0)
#define IS_TODO(it) (it->from == 0 && it->to == 0)
#define IS_EVENT(it) (!IS_TODO(it) && !IS_DEADLINE(it))

struct args {
	char *what;
	char *at;
	char *from;
	char *to;
};

struct item {
	char *what;
	char *at;
	time_t from;
	time_t to;
};

static void fail(struct args *a, const char *e, int show_usage);

static int read_args(struct args *a, const int argc, char *argv[]) {
	char c;

	while ((c = getopt(argc, argv, "w:a:f:t:")) != -1) {
		switch (c) {
			case 'w':
				a->what = malloc(strlen(optarg) + 1);
				strcpy(a->what, optarg);
				break;
			case 'a':
				a->at = malloc(strlen(optarg) + 1);
				strcpy(a->at, optarg);
				break;
			case 'f':
				a->from = malloc(strlen(optarg) + 1);
				strcpy(a->from, optarg);
				break;
			case 't':
				a->to = malloc(strlen(optarg) + 1);
				strcpy(a->to, optarg);
				break;
			case '?':
				return 0;
		}
	}

	return 1;
}

static void free_args(struct args *a) {
	free(a->what);
	free(a->at);
	free(a->from);
	free(a->to);
}

static void set_year_and_month(char *year, char *month, const struct tm *tm) {
	sprintf(year, "%d", 1900 + tm->tm_year);
	sprintf(month, "%02d", tm->tm_mon + 1);
}

/*
	For a short date format, if the user has typed in a day of the month
	that is before today, increase the month.
*/
static void inc_month(struct tm *tm, const char *s) {
	int day;
	char day_tmp[3];

	memset(day_tmp, 0, sizeof(day_tmp));

	if (s[0] == '0')
		memcpy(day_tmp, s, 1);
	else
		memcpy(day_tmp, s, 2);

	day = atoi(day_tmp);

	if (day < tm->tm_mday) {
		if (tm->tm_mon < 11)
			tm->tm_mon += 1;
		else {
			tm->tm_mon = 0;
			tm->tm_year++;
		}
	}
}

/*
	Add the current year or year and month to date string if the date specified
	is in a short format.
*/
static int complete_datestr(char *s1, const char *s2) {
	time_t t;
	struct tm *tm;
	char year[5];
	char month[3];

	time(&t);
	tm = localtime(&t);

	switch (strlen(s2)) {
		case 2:
		case 4:
		case 6:
			inc_month(tm, s2);
			set_year_and_month(year, month, tm);
			strcat(s1, year);
			strcat(s1, month);
			break;
		case 8:
			set_year_and_month(year, month, tm);
			strcat(s1, year);
			break;
		case 12:
			break;
		default:
			return 0;
	}

	strcat(s1, s2);

	switch (strlen(s2)) {
		case 2:
			strcat(s1, "00");
		case 4:
			strcat(s1, "00");
			break;
	}

	strcat(s1, "00");

	return 1;
}

static int datestr_to_time(time_t *t, const char *s) {
	struct tm tm;

	memset(&tm, 0, sizeof(struct tm));

	if (strptime(s, "%Y%m%d%H%M%S", &tm) == 0)
		return 0;
	else {
		tm.tm_isdst = -1;
		*t = mktime(&tm);

		return 1;
	}
}

static void get_filename(char *s) {
	sprintf(s, "%s/.f", getenv("HOME"));
}

static int write_to_file(struct args *a, const time_t from, const time_t to) {
	FILE *f;
	char fn[256];

	get_filename(fn);

	if ((f = fopen(fn, "a")) == NULL)
		return 0;

	if (a->what != 0)
		fprintf(f, "%s", a->what);

	fprintf(f, "\t");

	if (a->at != 0)
		fprintf(f, "%s", a->at);

	fprintf(f, "\t");

	if (from != 0)
		fprintf(f, "%llu", (unsigned long long)from);

	fprintf(f, "\t");

	if (to != 0)
		fprintf(f, "%llu", (unsigned long long)to);

	fprintf(f, "\n");
	fclose(f);

	return 1;
}

static void fail(struct args *a, const char *e, const int show_usage) {
	if (e != NULL)
		puts(e);

	if (show_usage)
		puts("Usage: f [-w what [-a at -f [[YYYY]MM]DD[hh[mm]] -t [[YYYY]MM]DD[hh[mm]]]]");

	free_args(a);

	exit(EXIT_FAILURE);
}

static int verify_args(const struct args *a) {
	if (a->what == 0)
		return 0;

	return 1;
}

static void list_items(const struct item* items) {
	int i;
	struct item it;
	struct tm *tm;
	char s[17];

	for (i = 0; i < LIST_ITEMS; i++) {
		it = items[i];

		if (it.what == NULL)
			break;

		printf("% 4d  ", i);

		if (!_IS_DEADLINE(it)) {
			if (it.from != 0) {
				tm = localtime(&it.from);
				strftime(s, sizeof(s), "%Y-%m-%d %H:%M", tm);
				printf("%s  ", s);
			}
		}
		else {
			tm = localtime(&it.to);
			strftime(s, sizeof(s), "%Y-%m-%d %H:%M", tm);
			printf("%sd ", s);
		}

		printf("%s", it.what);

		if (it.at != NULL)
			printf("@%s", items[i].at);

		printf("\n");

		if (!_IS_DEADLINE(it)) {
			if (it.to != 0) {
				tm = localtime(&it.to);
				strftime(s, sizeof(s), "%Y-%m-%d %H:%M", tm);
				printf("      %s\n", s);
			}
		}
	}
}

static void free_items(struct item* items) {
	int i;

	for (i = 0; i < LIST_ITEMS; i++) {
		if (items[i].what == 0)
			break;

		free(items[i].what);
		free(items[i].at);
	}
}

static void parse_read_item(struct item *it, char *line) {
	int col;
	char *token = NULL;
	char *delims = "\t";

	for (col = 0; (token = strsep(&line, delims)) != NULL; col++) {
		switch (col) {
			case 0:
				if (strlen(token) > 0) {
					it->what = malloc(strlen(token) + 1);
					strcpy(it->what, token);
				}

				break;
			case 1:
				if (strlen(token) > 0) {
					it->at = malloc(strlen(token) + 1);
					strcpy(it->at, token);
				}

				break;
			case 2:
				it->from = atoi(token);
				break;
			case 3:
				it->to = atoi(token);
				break;
		}
	}
}

static int read_items(struct item* items) {
	char fn[256];
	FILE *f;
	int n;
	ssize_t read;
	char *line = NULL;
	size_t len = 0;

	get_filename(fn);

	if ((f = fopen(fn, "r")) == NULL)
		return 0;

	for (n = 0; (read = getline(&line, &len, f)) != -1; n++) {
		line[read - 1] = '\0'; /* remove newline */
		parse_read_item(&items[n], line);
	}

	fclose(f);

	return n;
}

static int sort_items(const void *a, const void *b) {
	struct item *ia = (struct item *)a;
	struct item *ib = (struct item *)b;

	if (IS_EVENT(ia) && !IS_EVENT(ib))
		return -1;
	else if (IS_DEADLINE(ia) && !IS_DEADLINE(ib)) {
		if (IS_TODO(ib))
			return -1;
		else
			return 1;	
	}
	else if (IS_TODO(ia) && !IS_TODO(ib))
		return 1;
	else
		return ia->from > ib->from;

	return 0;
}

static int main_list_items() {
	int n;
	struct item *items;

	items = (struct item *)calloc(LIST_ITEMS, sizeof(struct item));

	n = read_items(items);
	qsort(items, n, sizeof(struct item), sort_items);
	list_items(items);
	free_items(items);

	return EXIT_SUCCESS;
}

int main_add_item(int argc, char *argv[]) {
	struct args a;
	char s[15];
	time_t from = 0;
	time_t to = 0;

	memset(&a, 0, sizeof(struct args));

	if (read_args(&a, argc, argv) == 0)
		fail(&a, "Could not parse arguments.", 1);

	if (verify_args(&a) == 0)
		fail(&a, NULL, 1);

	memset(&s, 0, sizeof(s));

	if (a.from != 0) {
		if (complete_datestr(s, a.from) == 0)
			fail(&a, "Could not complete from-date.", 1);

		if (datestr_to_time(&from, s) == 0)
			fail(&a, "Could not convert from-date string to time.", 1);
	}

	memset(&s, 0, sizeof(s));

	if (a.to != 0) {
		if (complete_datestr(s, a.to) == 0)
			fail(&a, "Could not complete to-date.", 1);

		if (datestr_to_time(&to, s) == 0)
			fail(&a, "Could not convert to-date string to time.", 1);
	}

	if (write_to_file(&a, from, to) == 0)
		fail(&a, "Could not write to ~/.f.", 0);

	free_args(&a);

	return main_list_items();
}

int main(int argc, char *argv[]) {
	if (argc < 2)
		return main_list_items();
	else
		return main_add_item(argc, argv);
}


flo.c
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>

#define LINE_LENGTH 1024
#define ITEM_COUNT 128
#define DATE_FORMAT "%Y-%m-%d %H:%M"
#define IS_DEADLINE(it) (it->from == 0 && it->to != 0)
#define IS_TODO(it) (it->from == 0 && it->to == 0)
#define IS_EVENT(it) (!IS_TODO(it) && !IS_DEADLINE(it))

struct args {
	char *what;
	char *at;
	char *from;
	char *to;
	int change;
	int remove;
	int id;
};

struct item {
	char *what;
	char *at;
	time_t from;
	time_t to;
};

int main(int argc, char *argv[]);
static int read_args(struct args *a, const int argc, char *argv[]);
static void read_args_short(struct args *a, const int argc, char *argv[]);
static void free_args(struct args *a);
static int list_items();
static int add_item(struct args *a);
static int change_item(struct args *a);
static int remove_item(struct args *a);
static int read_items(struct item *items);
static int write_items(const struct item *items, const int n, int except);
static int write_item(struct args *a, const time_t from, const time_t to);
static void print_items(const struct item *items, const int n);
static int sort_items(const void *a, const void *b);
static void free_items(struct item *items, const int n);
static int write_item_to_stream(
	FILE *f,
	const char *what,
	const char* at,
	time_t from,
	time_t to);
static void line_to_item(struct item *it, char *line);
static int parse_datestr(time_t *t, const char *s);
static int complete_datestr(char *s1, const char *s2);
static int datestr_to_time(time_t *t, const char *s);
static void inc_month(struct tm *tm, const char *s);
static void set_year_and_month(char *year, char *month, const struct tm *tm);
static int last_index_of(const char *s, const char c);
static void get_filename(char *s);
static void fail(struct args *a, const char *e, const int print_usage);

int main(int argc, char *argv[]) {
	struct args a;

	if (argc < 2)
		return list_items();
	else {
		memset(&a, 0, sizeof(struct args));

		if (argv[1][0] != '-') {
			read_args_short(&a, argc, argv);

			return add_item(&a);
		}
		else {
			if (read_args(&a, argc, argv) == 0)
				fail(&a, NULL, 1);

			if (a.change != 0)
				return change_item(&a);
			else if (a.remove != 0)
				return remove_item(&a);
			else
				return add_item(&a);
		}
	}
}

static int read_args(struct args *a, const int argc, char *argv[]) {
	char c;

	while ((c = getopt(argc, argv, "c:r:w:a:f:t:")) != -1) {
		switch (c) {
			case 'c':
				a->change = 1;
				a->id = atoi(optarg);
				break;
			case 'r':
				a->remove = 1;
				a->id = atoi(optarg);
				break;
			case 'w':
				a->what = malloc(strlen(optarg) + 1);
				strcpy(a->what, optarg);
				break;
			case 'a':
				a->at = malloc(strlen(optarg) + 1);
				strcpy(a->at, optarg);
				break;
			case 'f':
				a->from = malloc(strlen(optarg) + 1);
				strcpy(a->from, optarg);
				break;
			case 't':
				a->to = malloc(strlen(optarg) + 1);
				strcpy(a->to, optarg);
				break;
			case '?':
				return 0;
		}
	}

	return 1;
}

/*
	Parses a short version of the input string.
*/
static void read_args_short(struct args *a, const int argc, char *argv[]) {
	int i;
	char line[LINE_LENGTH];
	char *rest;
	int n;
	int len;

	memset(line, 0, sizeof(line));

	for (i = 1; i < argc; i++) {
		strcat(line, argv[i]);

		if (i != (argc - 1))
			strcat(line, " ");
	}

	n = last_index_of(line, '@');

	if (n != -1) {
		a->what = (char *)calloc(n + 1, 1);
		strncpy(a->what, line, n);
		rest = &line[0] + n + 1;
	}
	else
		rest = &line[0];

	n = last_index_of(rest, '-');

	if (n != -1) {
		len = strlen(rest) - n - 1;
		a->to = (char *)calloc(len, 1);	
		strncpy(a->to, rest + n + 1, len);
		rest[n] = '\0';
	}

	n = last_index_of(rest, ',');

	if (n != -1) {
		len = strlen(rest) - n - 1;
		a->from = (char *)calloc(len, 1);	
		strncpy(a->from, rest + n + 1, len);
		rest[n] = '\0';
	}

	if (a->what == 0) {
		a->what = (char *)calloc(strlen(rest) + 1, 1);
		strcpy(a->what, rest);	
	}
	else {
		a->at = (char *)calloc(strlen(rest) + 1, 1);
		strcpy(a->at, rest);	
	}
}

static void free_args(struct args *a) {
	free(a->what);
	free(a->at);
	free(a->from);
	free(a->to);
}

static int list_items() {
	int n;
	struct item *items;

	items = (struct item *)calloc(ITEM_COUNT, sizeof(struct item));

	n = read_items(items);
	qsort(items, n, sizeof(struct item), sort_items);
	print_items(items, n);

	free_items(items, n);

	return EXIT_SUCCESS;
}

static int add_item(struct args *a) {
	time_t from = 0;
	time_t to = 0;

	if (a->what == 0)
		fail(a, NULL, 1);

	if (a->from != 0)
		if (parse_datestr(&from, a->from) == 0)
			fail(a, "Could not parse from-date.", 1);

	if (a->to != 0)
		if (parse_datestr(&to, a->to) == 0)
			fail(a, "Could not parse to-date.", 1);

	if (write_item(a, from, to) == 0)
		fail(a, "Could not write to ~/.flo.", 0);

	free_args(a);

	return list_items();
}

static int change_item(struct args *a) {
	int n;
	struct item *items;
	struct item *it;

	items = (struct item *)calloc(ITEM_COUNT, sizeof(struct item));

	n = read_items(items);

	if (n == 0 || (a->id < 0 || a->id >= n)) {
		free_items(items, n);
		fail(a, "Could not find item.", 0);

		return EXIT_FAILURE;
	}

	qsort(items, n, sizeof(struct item), sort_items);

	it = &items[a->id];

	if (a->what != 0) {
		free(it->what);
		it->what = malloc(strlen(a->what) + 1);
		strcpy(it->what, a->what);
	}

	if (a->at != 0) {
		free(it->at);
		it->at = malloc(strlen(a->at) + 1);
		strcpy(it->at, a->at);
	}

	if (a->from != 0)
		if (parse_datestr(&it->from, a->from) == 0)
			fail(a, "Could not parse from-date.", 1);

	if (a->to != 0)
		if (parse_datestr(&it->to, a->to) == 0)
			fail(a, "Could not parse to-date.", 1);

	write_items(items, n, -1);
	
	free_args(a);
	free_items(items, n);

	return list_items();
}

static int remove_item(struct args *a) {
	int n;
	struct item *items;

	items = (struct item *)calloc(ITEM_COUNT, sizeof(struct item));

	n = read_items(items);

	if (n == 0 || (a->id < 0 || a->id > n)) {
		free_items(items, n);
		fail(a, "Could not find item.", 0);

		return EXIT_FAILURE;
	}

	qsort(items, n, sizeof(struct item), sort_items);

	write_items(items, n, a->id);

	free_args(a);
	free_items(items, n);

	return list_items();
}

static int read_items(struct item *items) {
	char fn[256];
	FILE *f;
	int n;
	char line[LINE_LENGTH];

	get_filename(fn);

	if ((f = fopen(fn, "r")) == NULL)
		return 0;

	for (n = 0; (fgets(line, LINE_LENGTH, f)) != NULL; n++) {
		line_to_item(&items[n], line);
	}

	fclose(f);

	return n;
}

static int write_items(const struct item *items, const int n, int except) {
	char fn[256];
	FILE *f;
	int i;

	get_filename(fn);

	if ((f = fopen(fn, "w")) == NULL)
		return 0;

	for (i = 0; i < n; i++) {
		if (i == except)
			continue;

		write_item_to_stream(
			f,
			items[i].what,
			items[i].at,
			items[i].from,
			items[i].to);
	}

	fclose(f);

	return 1;
}

static int write_item(struct args *a, const time_t from, const time_t to) {
	char fn[256];
	FILE *f;

	get_filename(fn);

	if ((f = fopen(fn, "a")) == NULL)
		return 0;

	write_item_to_stream(f, a->what, a->at, from, to);

	fclose(f);

	return 1;
}

static void print_items(const struct item *items, const int n) {
	int i;
	struct item *it;
	struct tm *tm;
	char s[17];

	for (i = 0; i < n; i++) {
		it = (struct item *)&items[i];

		if (IS_DEADLINE(it)) {
			tm = localtime(&it->to);
			strftime(s, sizeof(s), DATE_FORMAT, tm);
			printf("d% 3d  %s  %s", i, s, it->what);

			if (it->at != 0)
				printf("@%s", it->at);

			printf("\n");
		}
		else if (IS_TODO(it)) {
			printf("% 4d  %s", i, it->what);

			if (it->at != 0)
				printf("@%s", it->at);

			printf("\n");
		}
		else {
			tm = localtime(&it->from);
			strftime(s, sizeof(s), DATE_FORMAT, tm);
			printf("% 4d  %s  %s", i, s, it->what);

			if (it->at != 0)
				printf("@%s", it->at);

			printf("\n");

			if (it->to != 0) {
				tm = localtime(&it->to);
				strftime(s, sizeof(s), DATE_FORMAT, tm);
				printf("      %s\n", s);
			}
		}
	}
}

static int sort_items(const void *a, const void *b) {
	struct item *ia = (struct item *)a;
	struct item *ib = (struct item *)b;

	if (IS_EVENT(ia) && !IS_EVENT(ib))
		return -1;
	else if (IS_DEADLINE(ia) && !IS_DEADLINE(ib)) {
		if (IS_TODO(ib))
			return -1;
		else
			return ia->to > ib->from;
	}
	else if (IS_DEADLINE(ia) && IS_DEADLINE(ib))
		return ia->to > ib->to;
	else if (IS_TODO(ia) && !IS_TODO(ib))
		return 1;
	else
		return ia->from > ib->from;

	return 0;
}

static void free_items(struct item *items, const int n) {
	int i;

	for (i = 0; i < n; i++) {
		free(items[i].what);
		free(items[i].at);
	}

	free(items);
}

static int write_item_to_stream(
	FILE *f,
	const char *what,
	const char* at,
	time_t from,
	time_t to) {

	if (what != 0)
		fprintf(f, "%s", what);

	fprintf(f, "\t");

	if (at != 0)
		fprintf(f, "%s", at);

	fprintf(f, "\t");

	if (from != 0)
		fprintf(f, "%llu", (unsigned long long)from);

	fprintf(f, "\t");

	if (to != 0)
		fprintf(f, "%llu", (unsigned long long)to);

	fprintf(f, "\n");

	return 1;
}

static void line_to_item(struct item *it, char *line) {
	int col;
	char *token = NULL;
	char *delims = "\t";

	for (col = 0; (token = strsep(&line, delims)) != NULL; col++) {
		switch (col) {
			case 0:
				if (strlen(token) > 0) {
					it->what = malloc(strlen(token) + 1);
					strcpy(it->what, token);
				}

				break;
			case 1:
				if (strlen(token) > 0) {
					it->at = malloc(strlen(token) + 1);
					strcpy(it->at, token);
				}

				break;
			case 2:
				it->from = atoi(token);
				break;
			case 3:
				it->to = atoi(token);
				break;
		}
	}
}

static int parse_datestr(time_t *t, const char *s) {
	char s2[15];
	memset(s2, 0, sizeof(s2));

	if (complete_datestr(s2, s) == 0)
		return 0;

	if (datestr_to_time(t, s2) == 0)
		return 0;

	return 1;
}

/*
	Add the current year or year and month to date string if the date
	specified is in a short format.
*/
static int complete_datestr(char *s1, const char *s2) {
	time_t t;
	struct tm *tm;
	char year[5];
	char month[3];

	time(&t);
	tm = localtime(&t);

	switch (strlen(s2)) {
		case 2:
		case 4:
		case 6:
			inc_month(tm, s2);
			set_year_and_month(year, month, tm);
			strcat(s1, year);
			strcat(s1, month);
			break;
		case 8:
			set_year_and_month(year, month, tm);
			strcat(s1, year);
			break;
		case 12:
			break;
		default:
			return 0;
	}

	strcat(s1, s2);

	switch (strlen(s2)) {
		case 2:
			strcat(s1, "00");
		case 4:
			strcat(s1, "00");
			break;
	}

	strcat(s1, "00");

	return 1;
}

static int datestr_to_time(time_t *t, const char *s) {
	struct tm tm;

	memset(&tm, 0, sizeof(struct tm));

	if (strptime(s, "%Y%m%d%H%M%S", &tm) == 0)
		return 0;
	else {
		tm.tm_isdst = -1;
		*t = mktime(&tm);

		return 1;
	}
}

/*
	For a short date format, if the user has typed in a day of the month
	that is before today, increase the month.
*/
static void inc_month(struct tm *tm, const char *s) {
	int day;
	char day_tmp[3];

	memset(day_tmp, 0, sizeof(day_tmp));

	if (s[0] == '0')
		memcpy(day_tmp, s, 1);
	else
		memcpy(day_tmp, s, 2);

	day = atoi(day_tmp);

	if (day < tm->tm_mday) {
		if (tm->tm_mon < 11)
			tm->tm_mon += 1;
		else {
			tm->tm_mon = 0;
			tm->tm_year++;
		}
	}
}

static void set_year_and_month(char *year, char *month, const struct tm *tm) {
	sprintf(year, "%d", 1900 + tm->tm_year);
	sprintf(month, "%02d", tm->tm_mon + 1);
}

static int last_index_of(const char *s, const char c) {
	int i;

	for (i = strlen(s) - 1; i >= 0; i--) {
		if (s[i] == c)
			return i;
	}

	return -1;
}

static void get_filename(char *s) {
	sprintf(s, "%s/.flo", getenv("HOME"));
}

static void fail(struct args *a, const char *e, const int print_usage) {
	if (e != NULL)
		puts(e);

	if (print_usage)
    		puts("Usage: flo [what[@at][,from][-to] || [-c id] -w what | -\
a at | -f from | -t to || -r id]");

	free_args(a);

	exit(EXIT_FAILURE);
}
